### PR TITLE
EDGEPOD-428: s1ap-mme-app Interfaces changes to support dedicated bearer

### PR DESCRIPTION
**Pull request by clstefi**
This activity includes the following w.r.t dedicated bearer related messages.
1. S1AP/NAS Message Structures 
2. S1AP/NAS encoders and decoders 
3. S1AP handlers

### CODE REVIEW

**Comment by omecproject** [comment]
Can one of the admins verify this patch?

**Comment by omecproject** [comment]
Can one of the admins verify this patch?

**Comment by omecproject** [comment]
Can one of the admins verify this patch?

**Comment by omecproject** [comment]
Can one of the admins verify this patch?

File: include/common/msgType.h [hunk 1]
```diff
@@ -516,6 +543,18 @@ struct handover_cancel_ack_Q_msg {
...
 #define S1AP_HANDOVER_CANCEL_ACK_BUF_SIZE sizeof(struct handover_cancel_ack_Q_msg)
 
+struct erabsu_ctx_req_Q_msg {
+    msg_type_t msg_type;
+    uint32_t mme_ue_s1ap_id;
+    uint32_t enb_s1ap_ue_id;
+    ue_aggregate_maximum_bitrate ue_aggrt_max_bit_rate;
+    struct ERABSetupList erab_setup_list;
+    uint32_t enb_context_id;
+    Buffer nas_buf[MAX_ERAB_SIZE];
```
Comment (thakurajayL) [inline include/common/msgType.h:553]: What's this MAX_ERAB_SIZE..

File: include/common/nas_structs.h [hunk 2]
```diff
@@ -411,6 +411,86 @@ typedef struct tai_list {
 	struct TAI partial_list[EMM_MAX_TAI_LIST];
 } tai_list;
 
+/* 24.301 - 9.9.4.3.1 */
+typedef struct eps_qos_bitmap {
```
Comment (thakurajayL) [inline include/common/nas_structs.h:415]: structure name suggests bit map..but I see all are 1 byte length fields.

File: include/common/nas_structs.h [hunk 3]
```diff
@@ -411,6 +411,86 @@ typedef struct tai_list {
...
+} eps_qos_t;
+
+#define TFT_MAX_SIZE 256
+
+typedef struct bearer_tft{
+    uint16_t len;
+    uint8_t data[TFT_MAX_SIZE];
+} bearer_tft;
+
+typedef enum esm_cause_t {
```
Comment (thakurajayL) [inline include/common/nas_structs.h:443]: Would be good to have spec and section reference. That becomes handy for new folks

File: include/common/s1ap_structs.h [hunk 4]
```diff
@@ -436,6 +446,8 @@ typedef struct proto_IE_data {
         struct eRAB_elements 	erab;
         ue_aggregate_maximum_bitrate ue_aggrt_max_bit_rate;
         ERABSetup E_RABToBeSetupItemCtxtSUReq;
+	struct ERABSetupList eRABSetupList;
```
Comment (thakurajayL) [inline include/common/s1ap_structs.h:449]: Clearly no consistency in variable names... ue_aggregate_maximum_bitrate   , ERABSetup (typedefed).... struct ERABSetupList.... Please see if we can keep it consistent.

File: src/mme-app/utils/mmeNasUtils.cpp [hunk 5]
```diff
@@ -669,6 +669,98 @@ void MmeNasUtils::decode_tau_req(unsigned char *msg, int &nas_msg_len,
...
+            		datalen = msg[0];
+            		msg++;
+            		msg += datalen;
+            		break;
+		}
+		default:
+        	{
+            		log_msg(LOG_WARNING, "NAS ACT_DED_BR_CTXT_ACPT - Unhandled IE %x\n", elem_id);
+
+            		// set status to false to break out of while
```
Comment (thakurajayL) [inline src/mme-app/utils/mmeNasUtils.cpp:713]: We should skip unknown elements in NAS message payload. NAS payload comes from UE and we may have variety of UE equipments which various release support. So this piece of code need to be robust.

File: src/mme-app/utils/mmeNasUtils.cpp [hunk 6]
```diff
@@ -669,6 +669,98 @@ void MmeNasUtils::decode_tau_req(unsigned char *msg, int &nas_msg_len,
...
+                datalen = msg[0];
+                msg++;
+                msg += datalen;
+                break;
+            }
+            default:
+            {
+                log_msg(LOG_WARNING, "NAS ACT_DED_BR_CTXT_RJCT - Unhandled IE %x\n", elem_id);
+
+                // set status to false to break out of while
```
Comment (thakurajayL) [inline src/mme-app/utils/mmeNasUtils.cpp:755]: Skip the unknown IES/AVPs based on their length...else we would stuck in this while loop

File: src/mme-app/utils/mmeNasUtils.cpp [hunk 7]
```diff
@@ -669,6 +669,98 @@ void MmeNasUtils::decode_tau_req(unsigned char *msg, int &nas_msg_len,
...
+    uint8_t datalen = 0;
+    unsigned char *msg_end = msg + (nas_msg_len - ESM_HEADER_LEN);
+
+    nas->elements_len = 1;
+    nas->elements = (nas_pdu_elements *) calloc(sizeof(nas_pdu_elements), 1);
+    nas->elements[index].msgType = NAS_IE_TYPE_ESM_CAUSE;
+    memcpy(&nas->elements[index].pduElement.esm_msg.esm_cause, msg, 1);
+    msg++;
+    index++;
+#if 0
```
Comment (thakurajayL) [inline src/mme-app/utils/mmeNasUtils.cpp:734]: Is this. if 0 temporary or you plan to open this up later ?

File: src/mme-app/utils/mmeNasUtils.cpp [hunk 8]
```diff
@@ -2002,7 +2321,24 @@ void MmeNasUtils::encode_nas_msg(struct Buffer *nasBuffer, struct nasPDU *nas, S
...
+		{
+			log_msg(LOG_DEBUG, "Encoding Activate Dedicated Bearer Context Req NAS message in mme-app\n");
+			uint8_t mac_data_pos = MmeNasUtils::encode_act_ded_br_req(nasBuffer, nas);
+
+			/* Calculate mac */
+                        uint8_t direction = 1;
+                        uint8_t bearer = 0;
+                        unsigned char int_key[NAS_INT_KEY_SIZE];
+                        secContext.getIntKey(&int_key[0]);
+            		nas_int_algo_enum int_alg = secContext.getSelectIntAlg();
```
Comment (thakurajayL) [inline src/mme-app/utils/mmeNasUtils.cpp:2335]: Is this selecting new algorithm or just reading selecting new algorithm. For sure we can not change algorithm while creating dedicated bearers.

File: include/common/nas_structs.h (hunk 2 above)
Comment (clstefi) [reply include/common/nas_structs.h]: Moved the fields into struct eps_qos_t.

File: include/common/nas_structs.h (hunk 3 above)
Comment (clstefi) [reply include/common/nas_structs.h:440]: Taken care.

File: include/common/s1ap_structs.h (hunk 4 above)
Comment (clstefi) [reply include/common/s1ap_structs.h]: Done.

File: src/mme-app/utils/mmeNasUtils.cpp (hunk 5 above)
Comment (clstefi) [reply src/mme-app/utils/mmeNasUtils.cpp:713]: We have handled all IEs as per 23.401 v15.6.0

The length is not a mandatory field in a few of the NAS IEs, so we wont be able to fetch the length and skip the bytes unless we know which exact IE we are currently decoding.

If we keep incrementing the msg pointer (msg++) for default condition, we may encounter a byte that has same IE value as any of the valid IEs that we have handled, and may end up overwriting them. That's why we are stopping the decode once when we have encountered unhandled IE.

File: src/mme-app/utils/mmeNasUtils.cpp (hunk 6 above)
Comment (clstefi) [reply src/mme-app/utils/mmeNasUtils.cpp:755]: We have handled all IEs as per 23.401 v15.6.0

The length is not a mandatory field in a few of the NAS IEs, so we wont be able to fetch the length and skip the bytes unless we know which exact IE we are currently decoding.

File: src/mme-app/utils/mmeNasUtils.cpp (hunk 7 above)
Comment (clstefi) [reply src/mme-app/utils/mmeNasUtils.cpp:734]: For now, we are interested only in esm_cause. So, kept the optional IEs under #if 0 temporarily.

File: src/mme-app/utils/mmeNasUtils.cpp (hunk 8 above)
Comment (clstefi) [reply src/mme-app/utils/mmeNasUtils.cpp:2336]: We are just reading the selected algorithm. We aren't changing it here.

File: include/common/msgType.h [hunk 9]
```diff
@@ -259,6 +264,25 @@ struct s1apMsg_plus_raw_nas {
...
 }__attribute__ ((packed));
 
+struct erabSuResp_Q_msg {
+    int s1ap_enb_ue_id;
+    int s1ap_mme_ue_id;
+    erab_setup_list erab_su_list;
+    erab_failed_to_setup_list erab_fail_list;
+}__attribute__ ((packed));
+
+struct dedicatedBearerContextAccept_Q_msg {
```
Comment (thakurajayL) [inline include/common/msgType.h:274]: Should we move NAS messages to nas specific header file ? ...

Comment (thakurajayL) [inline include/common/msgType.h:274]:
```suggestion
/* Refer 36.413 - 9.1.3.2 */
struct dedicatedBearerContextAccept_Q_msg {
```

File: include/common/msgType.h [hunk 10]
```diff
@@ -259,6 +264,25 @@ struct s1apMsg_plus_raw_nas {
...
+    uint8_t eps_bearer_id;
+    uint8_t pti;
+    struct pco pco_opt;
+}__attribute__ ((packed));
+
+struct dedicatedBearerContextReject_Q_msg {
```
Comment (thakurajayL) [inline include/common/msgType.h:280]:
```suggestion
/* Refer 24.301  - 8.3.1.1  */
struct dedicatedBearerContextReject_Q_msg {
```

File: include/common/msgType.h [hunk 11]
```diff
@@ -259,6 +264,25 @@ struct s1apMsg_plus_raw_nas {
 	struct STMSI s_tmsi;
 }__attribute__ ((packed));
 
+struct erabSuResp_Q_msg {
```
Comment (thakurajayL) [inline include/common/msgType.h:267]:
```suggestion
/* Refer 36.413 - 9.1.3.2 */
struct erabSuResp_Q_msg {
```

File: include/common/msgType.h [hunk 12]
```diff
@@ -516,6 +543,18 @@ struct handover_cancel_ack_Q_msg {
 };
 #define S1AP_HANDOVER_CANCEL_ACK_BUF_SIZE sizeof(struct handover_cancel_ack_Q_msg)
 
+struct erabsu_ctx_req_Q_msg {
```
Comment (thakurajayL) [inline include/common/msgType.h:546]:
```suggestion
36.413 - 9.1.3.1
struct erabsu_ctx_req_Q_msg {
```

File: include/common/nas_headers.h [hunk 13]
```diff
@@ -57,6 +57,10 @@ extern "C"{
 #define NAS_DETACH_REQUEST 0x45
 #define NAS_DETACH_ACCEPT 0x46
 #define NAS_SERVICE_REQUEST 0x4D
+#define NAS_ACT_DED_BEARER_CTXT_ACPT 0xc6
```
Comment (thakurajayL) [inline include/common/nas_headers.h:60]:
```suggestion
// Refer - 24.301 , Table - 9.8.2  
#define NAS_ACT_DED_BEARER_CTXT_ACPT 0xc6
```

File: include/common/nas_headers.h [hunk 14]
```diff
@@ -57,6 +57,10 @@ extern "C"{
...
+#define NAS_ACT_DED_BEARER_CTXT_RJCT 0xc7
+
+#define ESM_HEADER_LEN 9
```
Comment (thakurajayL) [inline include/common/nas_headers.h:63]: I would be interested in knowing how we arrived at this 9.

File: include/common/nas_headers.h [hunk 15]
```diff
@@ -68,12 +72,13 @@ enum eps_nas_mesage_type {
...
-    TauReject    = 0x4b,
+    	TauReject    = 0x4b,
 	ServiceReject = 0x4e,
 	AuthenticationRequest = 0x52,
-    IdentityRequest       = 0x55,
+    	IdentityRequest       = 0x55,
 	SecurityModeCommand = 0x5d,
-    EMMInformation = 0x61,
+    	EMMInformation = 0x61,
+	ActivateDedicatedBearerContextRequest = 0xc5,
```
Comment (thakurajayL) [inline include/common/nas_headers.h:81]: Why we need #define NAS_ACT_DED_BEARER_CTXT_ACPT, if we already have eps_nas_message_type ?

File: include/common/nas_headers.h [hunk 16]
```diff
@@ -68,12 +72,13 @@ enum eps_nas_mesage_type {
 	DetachAccept = 0x46,
 	DetachRequest = 0x45,
 	TauAccept    = 0x49,
-    TauReject    = 0x4b,
+    	TauReject    = 0x4b,
```
Comment (thakurajayL) [inline include/common/nas_headers.h:75]: Can you add following at the start of this enum

/* Refer - 24.301 , Table - 9.8.1 */

File: include/common/msgType.h (hunk 9 above)
Comment (clstefi) [reply include/common/msgType.h:274]: Yes, msg structs can be moved to protocol specific header files. Shall we take that as a separate activity?

Comment (clstefi) [reply include/common/msgType.h:274]: Done.

File: include/common/msgType.h (hunk 10 above)
Comment (clstefi) [reply include/common/msgType.h:280]: Done.

File: include/common/msgType.h (hunk 11 above)
Comment (clstefi) [reply include/common/msgType.h:267]: Done.

File: include/common/msgType.h (hunk 12 above)
Comment (clstefi) [reply include/common/msgType.h:546]: Done.

File: include/common/nas_headers.h (hunk 13 above)
Comment (clstefi) [reply include/common/nas_headers.h:60]: Done.

File: include/common/nas_headers.h (hunk 14 above)
Comment (clstefi) [reply include/common/nas_headers.h:63]: NAS ESM Messages(Incoming) has the length of the NAS PDU Header equal to 9 Octets, distributed as below:
Security Header Type (1/2) + Protocol Discriminator (1/2) + MAC (4) + Sequence Number (1) +
EPS Bearer Identity (1/2) + Protocol Discriminator (1/2) + Procedure Transaction Identity (1) + NAS ESM Message Type (1)

File: include/common/nas_headers.h (hunk 15 above)
Comment (clstefi) [reply include/common/nas_headers.h:81]: In this enum, we are including outgoing NAS Messages (here, ActivateDedicatedBearerContextRequest).
In #define , we are including incoming NAS Messages (NAS_ACT_DED_BEARER_CTXT_ACPT & NAS_ACT_DED_BEARER_CTXT_RJCT).

File: include/common/nas_headers.h (hunk 16 above)
Comment (clstefi) [reply include/common/nas_headers.h:75]: Done.

File: include/common/msgType.h (hunk 9 above)
Comment (thakurajayL) [reply include/common/msgType.h:276]: Sure. No worry. if not big time effort do it now. I will review your code quickly to take to final stage quickly. We had lot of trouble in separating this code.

File: include/common/nas_headers.h (hunk 15 above)
Comment (thakurajayL) [reply include/common/nas_headers.h:90]: Well, I don't like it actually. Let me know if you agree to club them in single enum. I will do it, just want to know your opinion.

File: src/mme-app/utils/mmeS1MsgUtils.cpp [hunk 17]
```diff
@@ -58,11 +58,11 @@ void MmeS1MsgUtils::populateHoRequest(SM::ControlBlock& cb,
...
 
-    hoReq.eRABSetupList.count = 1;
-    hoReq.eRABSetupList.eRABSetup[0].e_RAB_ID = 5;
-    hoReq.eRABSetupList.eRABSetup[0].gtp_teid =
+    hoReq.erab_su_list.count = 1;
+    hoReq.erab_su_list.erab_su_item[0].e_RAB_ID = 5;
+    hoReq.erab_su_list.erab_su_item[0].gtp_teid =
             bearerCtxt->getS1uSgwUserFteid().fteid_m.header.teid_gre;
-    hoReq.eRABSetupList.eRABSetup[0].transportLayerAddress =
+    hoReq.erab_su_list.erab_su_item[0].transportLayerAddress =
```
Comment (thakurajayL) [inline src/mme-app/utils/mmeS1MsgUtils.cpp:65]: So we are currently not adding dedicated bearer in handover. Can we open jira ticket to handle handover for dedicated bearer later. @asreek

File: src/s1ap/handlers/attach_icsreq.c [hunk 18]
```diff
@@ -49,7 +49,7 @@ get_icsreq_protoie_value(struct proto_IE *value, struct init_ctx_req_Q_msg *g_ic
 	ieCnt++;
 
 	/* E-RABToBeSetupItemCtxtSUReq start */
-	ERABSetup *e_rab = &(value->data[ieCnt].val.E_RABToBeSetupItemCtxtSUReq);
+	erab_setup_item *e_rab = &(value->data[ieCnt].val.erab_to_be_setup_item_ctxt_su_req);
 	/* TODO: Remove hardcoded values. */
 	e_rab->e_RAB_ID = 1;
```
Comment (thakurajayL) [inline src/s1ap/handlers/attach_icsreq.c:54]: Point of discussion.. Radio Bearer Id need not be same as EPS bearer id ?

File: src/mme-app/utils/mmeS1MsgUtils.cpp (hunk 17 above)
Comment (asreek) [reply src/mme-app/utils/mmeS1MsgUtils.cpp:65]: EDGEPOD-436 is the story you had created to support HO with ded bearer feature. I was planning to take up this after we finish the ded bearer activation feature.

File: src/s1ap/handlers/attach_icsreq.c [hunk 19]
```diff
@@ -49,7 +49,7 @@ get_icsreq_protoie_value(struct proto_IE *value, struct init_ctx_req_Q_msg *g_ic
...
 	e_rab->e_RAB_QoS_Params.qci = 9;
```
Comment (thakurajayL) [inline src/s1ap/handlers/attach_icsreq.c:55]: Don't we get this from Mme-app ? What QCI to fill.

File: src/s1ap/handlers/erab_setup_response.c [hunk 20]
```diff
@@ -0,0 +1,136 @@
...
+void dump_erab_su_resp(struct erabSuResp_Q_msg *msg)
+{
+   log_msg(LOG_INFO, "MME-UE-S1AP-ID %d\n", msg->s1ap_mme_ue_id);
+   log_msg(LOG_INFO, "eNB-UE-S1AP-ID %d\n", msg->s1ap_enb_ue_id);
+   log_msg(LOG_INFO, "eRABSetupList_count %d\n", msg->erab_su_list.count);
+   for(int i = 0; i < msg->erab_su_list.count; i++)
+   {
+       log_msg(LOG_INFO,"eRAB_setup_item[%d].erab_id: %d\n", i, msg->erab_su_list.
+		       erab_su_item[i].e_RAB_ID);
+       log_msg(LOG_INFO,"eRAB_setup_item[%d].transportLayerAddress: %x\n", i, msg->erab_su_list.
```
Comment (thakurajayL) [inline src/s1ap/handlers/erab_setup_response.c:30]: to print address. Use inet_ntoa()

File: src/s1ap/handlers/erab_setup_response.c [hunk 21]
```diff
@@ -0,0 +1,136 @@
...
+{
+    s1_incoming_msg_data_t erab_su_resp = {0};
+    struct proto_IE erab_su_resp_ies = {0};
+
+    int decode_status = convertErabSetupRespToProtoIe(msg, &erab_su_resp_ies);
+    if (decode_status < 0)
+    {
+        log_msg(LOG_ERROR, "Failed to decode ERAB Setup Response\n");
+
+        if (erab_su_resp_ies.data != NULL)
```
Comment (thakurajayL) [inline src/s1ap/handlers/erab_setup_response.c:53]: even if its null, you can just call free.

File: src/s1ap/handlers/erab_setup_response.c [hunk 22]
```diff
@@ -0,0 +1,136 @@
...
+    //dump_erab_su_resp(&erab_su_resp.msg_data.erabSuResp_Q_msg_m);
+    int i = send_tipc_message(ipc_S1ap_Hndl, mmeAppInstanceNum_c,
+            (char*) &erab_su_resp,
+            S1_READ_MSG_BUF_SIZE);
+    if (i < 0)
+    {
+        log_msg(LOG_ERROR, "Error To write in erab_su_resp_handler %s\n", strerror(errno));
+    }
+
+    log_msg(LOG_INFO, "ERAB Setup Response sent to mme-app. Bytes sent %d\n", i);
```
Comment (thakurajayL) [inline src/s1ap/handlers/erab_setup_response.c:129]:
```suggestion else {
    log_msg(LOG_INFO, "ERAB Setup Response sent to mme-app. Bytes sent %d\n", i);
    } 
```

File: src/s1ap/handlers/erab_setup_response.c [hunk 23]
```diff
@@ -0,0 +1,136 @@
...
+
+    if (erab_su_resp_ies.data != NULL)
```
Comment (thakurajayL) [inline src/s1ap/handlers/erab_setup_response.c:131]:
```suggestion
```

**Comment by thakurajayL** [comment]
@yoooou ...why this is not triggering the Job ?
@badhrinathpa , we should merge Anajana's fix first or your  multiple eNB support first ?

**Comment by sindack** [comment]
add to whitelist

File: src/s1ap/handlers/attach_icsreq.c (hunk 18 above)
Comment (clstefi) [reply src/s1ap/handlers/attach_icsreq.c:54]: It needs to be same.

The value of the E-RAB ID used at S1 and X2 interfaces to identify an E-RAB allocated to the UE is the
same as the EPS Bearer ID value used at the Uu interface to identify the associated EPS Bearer (and also
used at the NAS layer as defined in TS 36.413 [25]). (36.300  sec:8.2)

File: src/s1ap/handlers/attach_icsreq.c (hunk 19 above)
Comment (clstefi) [reply src/s1ap/handlers/attach_icsreq.c:55]: This is the legacy code, that we retained from openmme. We don't get it from MME as of now.

File: src/s1ap/handlers/erab_setup_response.c (hunk 20 above)
Comment (clstefi) [reply src/s1ap/handlers/erab_setup_response.c:30]: Done.

File: src/s1ap/handlers/erab_setup_response.c (hunk 21 above)
Comment (clstefi) [reply src/s1ap/handlers/erab_setup_response.c:53]: Done.

File: src/s1ap/handlers/erab_setup_response.c (hunk 22 above)
Comment (clstefi) [reply src/s1ap/handlers/erab_setup_response.c:129]: Could you please clarify the difference between the suggested and the existing change?

File: src/s1ap/handlers/erab_setup_response.c (hunk 23 above)
Comment (clstefi) [reply src/s1ap/handlers/erab_setup_response.c:131]: Done.

File: include/common/nas_headers.h (hunk 15 above)
Comment (clstefi) [reply include/common/nas_headers.h:90]: Yes, we can club it . Changes done.

**Comment by thakurajayL** [comment]
@asreek  can you rebase your code ? I shall merge it after that,

**Review by thakurajayL — APPROVED** [review]

**Comment by thakurajayL** [comment]
@yoooou can you please help in knowing why this failed ?

**Comment by thakurajayL** [comment]
retest this please

**Pull request merged**